File: legacy_code/tries/no_prefix_set.cpp
#include <map>
#include <iostream>

using namespace std;

class TrieNode {
    private:
        char c;

    public:
        map<char, TrieNode*> children;

        TrieNode(char c) {
            this->c = c;
        }

        bool exist_child(char c) {
            if (children.find(c) != children.end()) return true;
            return false;
        }

        bool is_word() {
            return exist_child('*');
        }

        void add_child(char c) {
            if (!exist_child(c)) {
                children[c] = new TrieNode(c);
            }
        }

        void set_word() {
            add_child('*');
        }
};

class Trie {
    private:
        TrieNode* root;

    public:

        Trie() {
            root = new TrieNode('*');
        }

        void insert(string word) {
            TrieNode* current = root;
            for (int i = 0; i < word.size(); i++) {
                char c = word.at(i);
                current->add_child(c);
                current = current->children[c];
            }
            current->set_word();
        }

        bool exist_prefix(string word) {
            TrieNode* current = root;
            for (int i = 0; i < word.size(); i++) {
                char c = word.at(i);
                if (current->is_word()) return true;
                if (!current->exist_child(c)) return false;
                current = current->children[c];
            }
            return true;
        }
};


int main() {
    int n;
    Trie t;
    cin >> n;
    for(int i = 0; i < n; i++){
        string word;
        cin >> word;
        if (t.exist_prefix(word)){
            cout << "BAD SET" << endl;
            cout << word << endl;
            return 0;
        }
        t.insert(word);
    }
    cout << "GOOD SET" << endl;
    return 0;
}

File: legacy_code/tries/no_prefix_set.py
#!/bin/python
class Trie:

    def __init__(self):
        self.chars = {}

    def add(self, word):
        current = self.chars
        is_prefix = True
        for c in word:
            if c not in current:
                current[c] = {}
                is_prefix = False
            if "*" in current:
                return False
            current = current[c]
        if "*" not in current:
            current["*"] = None
            if is_prefix:
                return False
            else:
                return True
        else:
            return False

trie = Trie()

q = int(raw_input())
good = True
word = ""
for _ in xrange(q):
    word = raw_input()
    if not trie.add(word):
        good = False
        break

if good:
    print "GOOD SET"
else:
    print "BAD SET"
    print word
